### PR TITLE
several fixes for client's keep alive connections

### PR DESCRIPTION
**Pull request by CheSema**
Here are:
- client uses keep alive timeout from the server after negotiations 
- fix connection close race on client side
- fix logging 'Too many active sessions in group' is triggered extra times
- set default setting higher for DISK
- fix stored_count metric
- support keep alive max requests
- make default keep alive timeout for s3 as 5 seconds, aws s3 uses that value unchangeable 
- make default keep alive max request for s3 as 100, aws s3 uses that value unchangeable  

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Client has to send header 'Keep-Alive: timeout=X' to the server. If a client receives a response from the server with that header, client has to use the value from the server. Also for a client it is better not to use a connection which is nearly expired in order to avoid connection close race.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

File: src/Common/HTTPConnectionPool.cpp [hunk 1]
```diff
@@ -404,10 +420,18 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
...
 
         template <class... Args>
-        explicit PooledConnection(EndpointConnectionPool::WeakPtr pool_, IHTTPConnectionPoolForEndpoint::Metrics metrics_, Args &&... args)
-            : Session(args...), pool(std::move(pool_)), metrics(std::move(metrics_))
+        explicit PooledConnection(
+                EndpointConnectionPool::WeakPtr pool_,
+                ConnectionGroup::Ptr group_,
+                IHTTPConnectionPoolForEndpoint::Metrics metrics_,
+                Args &&... args)
+            : Session(args...)
```
Comment (nickitat) [inline src/Common/HTTPConnectionPool.cpp:428]: forward?

File: src/Common/HTTPConnectionPool.cpp [hunk 2]
```diff
@@ -213,7 +213,7 @@ class ConnectionGroup
 
         --total_connections_in_group;
 
-        const size_t reduced_warning_limit = limits.warning_limit > 10 ? limits.warning_limit - 10 : 1;
+        const size_t reduced_warning_limit = limits.warning_limit > 10 ? limits.warning_limit - 20 : 1;
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/HTTPConnectionPool.cpp:216]: It is a type, it is supposed to be the same val. Will fix.

File: src/Common/HTTPConnectionPool.cpp [hunk 3]
```diff
@@ -273,16 +273,24 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
...
-            ProfileEvents::increment(metrics.reset);
             Session::close();
 
             if (auto lock = pool.lock())
             {
                 auto timeouts = getTimeouts(*this);
                 auto new_connection = lock->getConnection(timeouts);
                 Session::assign(*new_connection);
+                if (Session::getKeepAliveRequest() == 0)
+                    Session::setKeepAliveRequest(1);
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/HTTPConnectionPool.cpp:293]: While I wes trying explain it. I understood that it should be just 
`Session::setKeepAliveRequest(Session::getKeepAliveRequest() + 1)`.
`reconnect()` is called in `sendRequst()` after `_keepAliveCurrentRequest += 1;` So we here are restore that `+1`.

File: src/Common/HTTPConnectionPool.cpp (hunk 1 above)
Comment (CheSema) [reply src/Common/HTTPConnectionPool.cpp:428]: You are right.

**Review by nickitat — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
A note for the record: the `stored_count` metric this PR fixed turned out to be the one counter that could have revealed a bug in the store limit, and it took until 26.9 for anything to consult it.

The fix here moved the metric update into a `SCOPE_EXIT` in `wipeExpiredImpl`, so `DiskConnectionsStored` became accurate on the expiry path as well:

```cpp
SCOPE_EXIT({
    CurrentMetrics::sub(getMetrics().stored_count, expired_connections.size());
    ProfileEvents::increment(getMetrics().expired, expired_connections.size());
});
```

The metric was correct after that, but it was only a metric. The decision whether to cache a connection went through `isStoreLimitReached`, which compared `total_connections_in_group` — active plus stored — against `store_limit`. So the number that described the cache and the number that governed the cache were different, and they diverged exactly when it mattered: on a busy server the total sat far above the limit while the stored count sat at 0, and the pool cached nothing at all. A production case on 26.8.2 showed `DiskConnectionsStored = 0` next to `DiskConnectionsTotal = 11,987`, with 99.2% of connections reset after one request and the host's ephemeral port range exhausted twice over (https://github.com/ClickHouse/ClickHouse/issues/119072).

https://github.com/ClickHouse/ClickHouse/pull/119078 makes the two the same counter: the group keeps `stored_connections_in_group`, the limit is checked against it, and the metric is maintained by the same reserve/release pair, so `DiskConnectionsStored` can no longer disagree with the limit that produced it.

**Pull request merged**